_posts/got-the-time.md

```diff
@@ -0,0 +1,244 @@
+
+> For more information on messages correlation, refer to the [NServiceBus documentation](https://docs.particular.net/nservicebus/sagas/message-correlation).
+
+One could argue that if the saga can be started by the `InvoiceIssued` even, there is no point in mapping the message as, by definition, there won't be an existing saga for that invoice. There at least a couple of cases in which we need that:
+
+- Duplicate messages: if we're dealing with an "at-least-once" type of transport, we should expect that the infrastructure can deliver the same message multiple times. In that case, we don't want to create a saga instance for each duplicate message; instead, we want to dispatch duplicates to the same saga instance and handle them idempotently.
+- Out of order messages: There are scenarios in which sagas can have many entry points, e.g., more than one message type can start a saga. If this is the case, we cannot assume that the `InvoiceIssued` event is the first to arrive. The saga instance might have been already created by a different message.
+
+Back to the overdue invoices tracking scenario; the second condition is: a paid invoice is by definition not overdue. We can model that by handling the `InvocePaid` event:
+
+```csharp
+class OverdueInvoicePolicy :
+   Saga<OverdueInvoicePolicy.OverdueInvoiceData>,
+   IAmStartedByMessages<InvoiceIssued>,
+   IHandleMessages<InvoicePaid>
+{
+   public Task Handle(InvoicePaid message, IMessageHandlerContext context)
+   {
+      MarkAsComplete();
+      return Task.CompletedTask;
+   }
+}
+```
+
+(rest of the code omitted for clarity)
+
+Two things worth noting:
+
+1. `InvoicePaid` is marked as `IHandleMessages<>`, which means that the `InvoicePaid` event cannot start the saga.
+2. When the invoice is paid, we can mark the saga as complete; as said, a paid invoice is by definition not overdue; we don't need to keep track of it anymore.
+
+## Delayed deliveries to model time
+
+The initial skeleton of the process is in place. It's now time to design the overdue check. In our case, all issued invoices are due in thirty days. We assume that a bank wire transfer might take a few days to appear on our bank account. Given the two conditions, it's worth checking invoice payments after forty days. When using sagas, we can leverage saga timeouts as a handy tool to model time:
+
+```csharp
+public async Task Handle(InvoiceIssued message, IMessageHandlerContext context)
+{
+   var checkDate = DateTime.Now.AddDays(40);
+   await RequestTimeout<CheckPayment>(context, checkDate);
+}
+```
+
+> The usage of `DateTime.Now` is terrible. It makes the code untestable and not timezones safe. Production-ready code needs to inject a component that can provide that information and use `DateTimeOffset` and account for time zones.
```

Comment:
Is it worth mentioning the magic number of 40 as well? You could say the whole thing should be offloaded to a `GetPaymentPolicyDueDate` component method but since this is an aside, I won't argue if you want to keep it simple.